### ISSUE TITLE
Question about uneven distribution of GPU memory

### ISSUE THREAD

**Issue by Xiaolong-RRL** [BAAI-DCAI/Bunny#33]
Dear author:

Thanks for your interesting work.

During the full or lora finetuning, the memory usage of different GPUs is uneven：
![image](https://github.com/BAAI-DCAI/Bunny/assets/74476562/037c1c85-c4b1-42e1-99e8-b8b258c9e943)

I wander if it's correct? And how to handle it?

Thanks!!

**Comment by Jancsi9981** [comment]
I also encountered the same problem, do you know how to solve it now?

**Comment by GewelsJI** [comment]
Hi, @RussRobin 
Thanks for your detailed illustration here.

Actually, you've set the hyper-para `model_max_length=2048`, so that the whole bunch of input tokens would be the same, right?

Another question is: does the length of supervised sentence affect the GPU memory allocation?

Best,
Daniel

**Comment by RussRobin** [comment]
TL;DR: No. Yes.

Max token = 2048 sets the upper bound of token number, and the upper bound of GPU memory.

Here is how the tokenizer works for phi-2 and siglip with our default parameters: an image is encoded into 792 tokens, and all texts in QA pairs are encoded ( in most cases a word means 1-2 tokens). So the token length for a sample is related to the length of QA. For most samples in Bunny_695k, token length is no more than 2048.

GPU allocation is related to the actual token length, and should be no more than the max token length.

You can try: in Bunny_695k, set max token length to 4096, and the GPU memory won’t change too much as compared to 2048.

Feel free to reach out if you have further question!


Regards
Russell
BAAI

**Comment by Isaachhh** [comment]
The distribution of GPU memory if training Bunny on a dataset of all same samples:
![Screenshot 2024-04-17 at 10 31 27](https://github.com/BAAI-DCAI/Bunny/assets/93458689/f1b1720a-e9f0-45fb-8a40-f9110663b831)

**Comment by GewelsJI** [comment]
> The distribution of GPU memory if training Bunny on a dataset of all same samples: ![Screenshot 2024-04-17 at 10 31 27](https://private-user-images.githubusercontent.com/93458689/323052898-f1b1720a-e9f0-45fb-8a40-f9110663b831.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTMzMjk5NDgsIm5iZiI6MTcxMzMyOTY0OCwicGF0aCI6Ii85MzQ1ODY4OS8zMjMwNTI4OTgtZjFiMTcyMGEtZTlmMC00NWZiLThhNDAtZjkxMTA2NjNiODMxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDE3VDA0NTQwOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTY4ZGNjZGU4ZGQwMmM0M2NiYTYzNzNjMmZjODI5YjczN2E0ZjZmNTE1MWNjN2FiZjZjYmJhMTczMTdkYzM2ZGImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.I-3xN0KcvgrKUXZb3UJSOKxNX7n23mvkyINRsq2KggQ)

How much batch size did you set in this test?

**Comment by GewelsJI** [comment]
> TL;DR: No. Yes.
> 
> Max token = 2048 sets the upper bound of token number, and the upper bound of GPU memory.
> 
> Here is how the tokenizer works for phi-2 and siglip with our default parameters: an image is encoded into 729 tokens, and all texts in QA pairs are encoded ( in most cases a word means 1-2 tokens). So the token length for a sample is related to the length of QA. For most samples in Bunny_695k, token length is no more than 2048.
> 
> GPU allocation is related to the actual token length, and should be no more than the max token length.
> 
> You can try: in Bunny_695k, set max token length to 4096, and the GPU memory won’t change too much as compared to 2048.
> 
> Feel free to reach out if you have further question!
> 
> Regards Russell BAAI

Appreciated to your detailed response here. Thanks.

**Comment by Isaachhh** [comment]
> How much batch size did you set in this test?

8 in each GPU, as default in the training script

**Comment by GewelsJI** [comment]
Awesome, thank you guys. 

Cheers,
Daniel.

**Comment by Xiaolong-RRL** [comment]
> Hi @Xiaolong-RRL & @Jancsi9981 ,
> 
> Thank you for your interest in our work.
> 
> Yes, uneven distribution of GPU memory is expected. Don’t worry about it. The reason behind is quite straight forward: Say, per device batch size is 8 and each image is encoded to 729 tokens. However, the text length for each sample is different, and will be encoded to different numbers of tokens, thus memory consumption on each GPU varies.
> 
> A side verification: when I train Bunny on one dataset, where word number in each QA is almost identical, memory consumption on each GPU is almost the same.
> 
> You may wonder why doesn’t GPU consumption remains almost the same (and imbalanced) during the whole training process. This is because of the memory reservation policy of torch. When memory is reserved, it won’t be freed even if it’s not used in the future, until the subprocess is killed. Let’s say in most cases 40G memory is needed. But some samples on GPU:0 in epoch 1/1000 are super long, so torch reserves 70G on GPU:0, and will not free unused 30G in the following steps. So nvidia-smi always show you unbalanced GPU memory consumption.
> 
> Please feel free to further comment on this issue if you still get confused.
> 
> Regards Russell BAAI

Thanks for your detailed and kindly reply!

**Comment by RussRobin** [comment]
I'll close this issue since it seems that we have reached a consensus on it. Thank you again for your interest in our work.
